nengo/utils/sockets.py

```diff
@@ -0,0 +1,322 @@
+from __future__ import absolute_import
+
+import struct
+import socket
+import time
+import threading
+import errno
+import warnings
+
+from nengo.utils.compat import queue
+
+
+class SocketCheckAliveThread(threading.Thread):
+    def __init__(self, socket_class):
+        threading.Thread.__init__(self)
+        self.socket_class = socket_class
+
+    def run(self):
+        # Keep checking if the socket class is still being used.
+        while (time.time() - self.socket_class.last_active <
+               self.socket_class.max_idle_time_current):
+            time.sleep(self.socket_class.alive_thread_sleep_time)
+        # If the socket class is idle, terminate the sockets
+        self.socket_class._close_recv_socket()
+        self.socket_class._close_send_socket()
+
+
+class UDPSocket:
+    def __init__(self, send_dim=1, recv_dim=1, dt_remote=0,
+                 local_port=-1, dest_addr='127.0.0.1', dest_port=-1,
+                 timeout=30, max_idle_time=1):
+        self.local_addr = '127.0.0.1'
+        self.local_port = local_port
+        self.dest_addr = (dest_addr if isinstance(dest_addr, list)
+                          else [dest_addr])
+        self.dest_port = (dest_port if isinstance(dest_port, list)
+                          else [dest_port])
+        self.timeout = timeout
+        self.byte_order = '!'
+
+        self.last_t = 0.0
+        self.dt = 0.0                               # Local simulation dt
+        self.dt_remote = max(dt_remote, self.dt)    # dt btw each packet sent
+        self.last_packet_t = 0.0
+
+        self.last_active = time.time()
+        self.max_idle_time_initial = max_idle_time
+        self.max_idle_time_timeout = max(max_idle_time, timeout + 1)
+        self.max_idle_time_current = self.max_idle_time_timeout
+        self.alive_check_thread = None
+        self.alive_thread_sleep_time = max_idle_time / 2.0
+        self.retry_backoff_time = 1
+
+        self.send_socket = None
+        self.recv_socket = None
+        self.is_sender = dest_port != -1
+        self.is_receiver = local_port != -1
+        self.ignore_timestamp = False
+
+        self.send_dim = send_dim
+        self.recv_dim = recv_dim
+
+        self.max_recv_len = (recv_dim + 1) * 4
+        self.value = [0.0] * recv_dim
+        self.buffer = queue.PriorityQueue()
+
+    def __del__(self):
+        self.close()
+
+    def _initialize(self):
+        self.value = [0.0] * self.recv_dim
+        self.last_t = 0.0
+        self.last_packet_t = 0.0
+
+        # Empty the buffer
+        while (not self.buffer.empty()):
+            self.buffer.get()
+
+    def _open_socket(self):
+        # Close socket, terminate alive check thread
+        self.close()
```

Comment:
Why do this at every time step? It seems kind of inefficient?

Reply:
this is only called once as threads are instantiated